### ISSUE TITLE
feat: Coarse-to-Fine Hierarchical Flood Fill Engine (sign_mode=5)

### ISSUE THREAD

**Issue by KhoiDOO** [KhoiDOO/conquer3d#19]
## Overview
Implementation and integration of the Unified 2-Level Multi-Scale Graph Wavefront Propagation engine for hierarchical volumetric flood fill (`sign_mode=5`). Reduces peak GPU memory consumption by up to **$314.7\times$** while maintaining **$100.000\%$ exact topological and geometric parity** with dense single-level flood fill.

**Comment by KhoiDOO** [comment]
### Technical Implementation Documentation & Architectural Audit: Coarse-to-Fine Flood Fill (`sign_mode=5`)

---

### 1. API Usage & Executable Python Code Snippets

```python
import torch
import conquer3d
from conquer3d.data.assets import Armadillo

# 1. Load 3D Triangle Mesh Asset (Armadillo)
mesh = Armadillo(device="cuda")

# 2. Build Coarse-to-Fine Flood Fill Data Structure
# Automatically computes optimal macro-block divisor (e.g. 8x8x8) and coarse/fine masks
block_size = [8, 8, 8]
grid_res = [512, 512, 512]
cf_data = mesh.build_flood_fill_cf_data(grid_res=grid_res, block_size=block_size)

# 3. High-Performance Point Query with sign_mode=5
# Evaluates signed distance values with O(1) early-exit on exterior/interior macro-blocks
query_pts = torch.rand((1000000, 3), device="cuda") * 2.0 - 1.0
sdf_values = mesh.query_points(query_pts, sign_mode=5)

# 4. End-to-End Watertight Isosurface Extraction with tmesh2sparse
# Converts mesh directly to sparse voxel grid using sign_mode=5
sparse_grid = conquer3d.conversion.tmesh2sparse(
    mesh,
    grid_res=512,
    sign_mode=5,
    block_size=[8, 8, 8]
)

# 5. Extract Watertight Meshes using any Extraction Algorithm
verts_mc, faces_mc = conquer3d.ops.marching_cubes(sparse_grid.coords, sparse_grid.values)
verts_dc, faces_dc = conquer3d.ops.dual_contouring(sparse_grid.coords, sparse_grid.values)
verts_dmc, faces_dmc = conquer3d.ops.dual_marching_cubes(sparse_grid.coords, sparse_grid.values)
```

---

### 2. Advantages & Core Capabilities
- **Massive VRAM Reduction**: Drops peak VRAM allocation from dense full-grid int8/int32 buffers down to a compact 2-level representation:
  - $512^3$ resolution: **$5.58\text{ MB}$ vs $1152.00\text{ MB}$ ($206.6\times$ reduction)**.
  - $1024^3$ resolution: **$29.28\text{ MB}$ vs $9216.00\text{ MB}$ ($314.7\times$ reduction)** — saving over **$9.18\text{ GB}$ of VRAM** on consumer GPUs.
- **$100.000\%$ Topological Parity**: Unified 2-level graph propagation guarantees identical vertex counts, face counts, and watertight topology compared to dense flood fill.
- **$O(1)$ Early Exit Point Queries**: BVH point queries evaluate coarse macro-blocks first; voxels in open air ($+1$) or deep solid ($-1$) resolve in 1 memory lookup without traversing BVH leaf triangles.
- **Dynamic Block Divisibility**: Automatic divisor calculation ensures arbitrary non-power-of-two grid resolutions seamlessly divide into valid block configurations.

---

### 3. Disadvantages & Known Trade-offs
- **Iterative Wavefront Convergence**: Unified multi-scale propagation runs synchronized host loop iterations across coarse and fine kernels until convergence, introducing slight kernel launch overhead compared to single-pass uncompressed operations.
- **Shared Memory Limit per Block**: Fine intra-block BFS uses `int8_t s_mask[8][8][8]` ($512$ bytes shared memory per CUDA threadblock), capping the ideal macro-block dimension to $8\times 8\times 8$ per block.

---

### 4. Empirical Benchmarks & Performance Metrics

#### Benchmark Asset: Stanford Armadillo ($49,990$ vertices, $99,976$ faces) on NVIDIA RTX GPU

| Resolution | Grid Voxels | Dense VRAM (`sign_mode=3`) | CF VRAM (`sign_mode=5`) | Memory Reduction | SDF Sign Parity |
| :--- | :--- | :--- | :--- | :---: | :---: |
| **$512^3$** | $134,217,728$ | $1,152.00\text{ MB}$ | **$5.58\text{ MB}$** | **$206.6\times$** | **$100.000\%$** |
| **$1024^3$** | $1,073,741,824$ | $9,216.00\text{ MB}$ | **$29.28\text{ MB}$** | **$314.7\times$** | **$100.000\%$** |

#### Isosurface Extraction Parity (Zero Missing Vertices / Zero Holes)

| Method | Resolution $512^3$ Vertices | Resolution $512^3$ Faces | Resolution $1024^3$ Vertices | Resolution $1024^3$ Faces | Parity Status |
| :--- | :--- | :--- | :--- | :--- | :---: |
| **Marching Cubes (MC)** | $466,196$ | $932,388$ | $1,872,248$ | $3,744,492$ | **100% Identical** |
| **Dual Contouring (DC)** | $466,174$ | $932,392$ | $1,872,220$ | $3,744,496$ | **100% Identical** |
| **Dual Marching Cubes (DMC)** | $466,198$ | $932,392$ | $1,872,250$ | $3,744,496$ | **100% Identical** |
| **Marching Tetrahedra (MT)** | $1,421,782$ | $2,843,560$ | — | — | **100% Identical** |

---

### 5. Technical Insights & Architectural Decisions
- **Unified 2-Level Multi-Scale Graph Wavefront**:
  - `Coarse-to-Coarse`: Expands exterior water across empty macro-blocks.
  - `Coarse-to-Fine`: Seeds boundary block faces touching coarse exterior water via BVH line-of-sight tests.
  - `Fine Intra-Block BFS`: Propagates water in shared memory avoiding SAT-colliding triangle boundaries.
  - `Fine-to-Fine Inter-Block Exchange`: Exchanging water across adjacent boundary block faces.
  - `Fine-to-Coarse`: Seeds unreached empty coarse blocks from fine water voxels.
- **Exact Separating Axis Theorem (SAT)**: Macro-blocks are filtered with exact 3D triangle-box intersection tests (`Triangle::is_voxel_intersect`), preventing false boundary classification.
